packages/core/src/bundle.ts

```diff
@@ -27,15 +40,164 @@ export function convertToTransactionBundle(bundle: Bundle): Bundle {
         resource: entry.resource,
       })),
     },
-    replacer,
+    (key, value) => referenceReplacer(key, value, idToUuid),
     2
   );
-  return JSON.parse(jsonString);
+  return reorderBundle(JSON.parse(jsonString) as Bundle);
 }
 
-function replacer(key: string, value: string): string {
+function referenceReplacer(key: string, value: string, idToUuid: Record<string, string>): string {
   if (key === 'reference' && typeof value === 'string' && value.includes('/')) {
-    return 'urn:uuid:' + value.split('/')[1];
+    const id = value.split('/')[1];
+    return 'urn:uuid:' + idToUuid[id];
   }
   return value;
 }
+
+/**
+ * Topologically sorts a `batch` or `transaction` bundle to improve reference resolution.
+ * The bundle is sorted such that a resource is created _before_ references to that resource appear in the bundle.
+ *
+ * In the event of cycles, this function will first create a POST request for each resource in the cycle, and then will
+ * append a PUT request to the bundle. This ensures that each resources in the cycle is visited twice, and all
+ * references can be resolved
+ * @param bundle Input bundle with type `batch` or `transaction`
+ * @returns Bundle of the same type, with Bundle.entry reordered
+ */
+export function reorderBundle(bundle: Bundle): Bundle {
```

Comment:
Ohhh boy.... Let's discuss this offline.  We've tried this in the past.  I'm slightly against adding this.

Reply:
K, discussed offline, let's do it 👍